ss/pebbledb/db.go

```diff
@@ -186,13 +187,84 @@ func (db *Database) ApplyChangeset(version int64, cs *proto.NamedChangeSet) erro
 	return b.Write()
 }
 
-// Prune for the PebbleDB SS backend is currently not supported. It seems the only
-// reliable way to prune is to iterate over the desired domain and either manually
-// tombstone or delete. Either way, the operation would be timely.
-//
-// See: https://github.com/cockroachdb/cockroach/blob/33623e3ee420174a4fd3226d1284b03f0e3caaac/pkg/storage/mvcc.go#L3182
+// Prune attempts to prune all versions up to and including the current version
+// Get the range of keys, manually iterate over them and delete them
 func (db *Database) Prune(version int64) error {
-	panic("not implemented!")
+	itr, err := db.storage.NewIter(nil)
+	if err != nil {
+		return err
+	}
+	defer itr.Close()
+
+	batch := db.storage.NewBatch()
+	var counter int
+	var prevKey, prevKeyEncoded []byte
+	var prevVersionDecoded int64
+
+	for itr.First(); itr.Valid(); {
+		// Store current key and version
+		currKeyEncoded := slices.Clone(itr.Key())
+		currKey, currVersion, currOK := SplitMVCCKey(currKeyEncoded)
+		if !currOK {
+			return fmt.Errorf("invalid MVCC key")
+		}
+
+		currVersionDecoded, err := decodeUint64Ascending(currVersion)
+		if err != nil {
+			return err
+		}
+
+		// Seek to next key if we are at a version which is higher than prune height
+		if currVersionDecoded > version {
+			itr.NextPrefix()
+			continue
+		}
+
+		// Deletes a key if another entry for that key exists a larger version than original but leq prune height
+		if bytes.Equal(prevKey, currKey) && prevVersionDecoded <= version && currVersionDecoded <= version {
+			// Delete previous key
+			err = batch.Delete(prevKeyEncoded, defaultWriteOpts)
+			if err != nil {
+				return err
+			}
+
+			counter++
+			if counter >= PruneCommitBatchSize {
+				err = batch.Commit(defaultWriteOpts)
+				if err != nil {
+					return err
+				}
+				err = batch.Close()
+				if err != nil {
+					return err
+				}
+
+				counter = 0
+				batch = db.storage.NewBatch()
+			}
+		}
+
+		// Update prevKey and prevVersion for next iteration
+		prevKey = currKey
+		prevVersionDecoded = currVersionDecoded
+		prevKeyEncoded = currKeyEncoded
+
+		itr.Next()
+	}
+
+	// Commit any leftover delete ops in batch
+	if counter > 0 {
+		err = batch.Commit(defaultWriteOpts)
+		if err != nil {
+			return err
+		}
+		err = batch.Close()
+		if err != nil {
+			return err
```

Comment:
I would defer close the batch right after you create it above.

Reply:
(I agree with this)